File: backend/app/api/admin/production/work_orders.py
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from app.core.deps import get_current_user, get_db, require_permissions
from app.core.response import ok
from app.crud.work_order import get_work_order_by_id, list_work_orders
from app.models.sku import Sku
from app.models.tenant import Tenant
from app.models.user import User
from app.models.work_order import WorkOrder
from app.services.entity_refs import process_ref_dict, product_ref_dict, sku_ref_dict


router = APIRouter(dependencies=[Depends(require_permissions(["work.manage"]))])


def _task_out(x) -> dict:
    return {
        "id": x.id,
        "tenant_id": x.tenant_id,
        "work_order_id": x.work_order_id,
        "process_id": x.process_id,
        "seq": x.seq,
        "planned_qty": x.planned_qty,
        "status": x.status,
        "assigned_user_id": x.assigned_user_id,
        "assigned_at": x.assigned_at,
        "assigned_by": x.assigned_by,
        "created_at": x.created_at,
        "updated_at": x.updated_at,
        "process": process_ref_dict(x.process),
    }


def _out(x) -> dict:
    sku = x.sku if hasattr(x, "sku") else None
    product = x.product if hasattr(x, "product") else None
    sku_out = sku_ref_dict(sku, product)
    return {
        "id": x.id,
        "tenant_id": x.tenant_id,
        "order_id": x.order_id,
        "order_item_id": x.order_item_id,
        "product_id": x.product_id,
        "sku_id": x.sku_id,
        "qty": x.qty,
        "status": x.status,
        "created_at": x.created_at,
        "updated_at": x.updated_at,
        "product": product_ref_dict(product),
        "sku": sku_out,
        "sku_display_label": sku_out["display_label"] if sku_out else None,
    }


@router.get("")
def list_api(
    order_id: int | None = Query(default=None, ge=1),
    status: str | None = Query(default=None),
    offset: int = Query(default=0, ge=0),
    limit: int = Query(default=50, ge=1, le=200),
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    items = list_work_orders(db, tenant_id=user.tenant_id, order_id=order_id, status=status, offset=offset, limit=limit)
    return ok({"items": [_out(x) for x in items]})


@router.get("/{work_order_id}")
def get_api(
    work_order_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    item = get_work_order_by_id(db, tenant_id=user.tenant_id, work_order_id=work_order_id, with_tasks=True)
    if not item:
        raise HTTPException(status_code=400, detail="工单不存在")
    data = _out(item)
    data["tasks"] = [_task_out(x) for x in item.tasks]
    return ok(data)

File: frontend-admin-pro/src/api/work-orders.ts
import { http } from '@/utils/http'
import type { ListResp } from '@/types/api'

export type WorkOrderOut = {
  id: number
  tenant_id: number
  order_id: number
  order_item_id: number
  product_id: number
  sku_id: number
  qty: number
  status: string
  created_at: string
  updated_at: string
  sku: {
    id: number
    code: string
    name: string
    product_id?: number
    product_name?: string
    sku_name?: string
    display_label?: string
  } | null
}

export type WorkOrderDetailOut = WorkOrderOut & {
  tasks: TaskOut[]
}

export type TaskAssignmentOut = {
  id: number
  user_id: number
  assigned_qty: number
  reported_qty?: number
  remaining_qty?: number
  assigned_at: string
  assigned_by: number | null
  user: { id: number; username: string; full_name: string | null } | null
}

/** 分工分配列表行（员工×任务） */
export type DispatchAssignmentOut = {
  id: number
  task_id: number
  task_code: string
  task_status: string
  order_id: number | null
  order_code: string | null
  product_id: number | null
  product_name: string | null
  sku_id: number | null
  sku_code: string | null
  sku_name: string | null
  process_id: number | null
  process_code: string | null
  process_name: string | null
  user_id: number
  username: string | null
  user_full_name: string | null
  assigned_qty: number
  reported_qty: number
  remaining_qty: number
  progress: number
  status: string
  assigned_at: string
  assigned_by: number | null
}

export type TaskOut = {
  id: number
  tenant_id: number
  work_order_id: number
  process_id: number
  seq: number
  task_code: string
  planned_qty: number
  status: string
  assigned_user_id: number | null
  assigned_at: string | null
  assigned_by: number | null
  equipment_id?: number | null
  assignments?: TaskAssignmentOut[]
  assigned_total_qty?: number
  unassigned_qty?: number
  created_at: string
  updated_at: string
  process: { id: number; code: string; name: string } | null
  equipment?: { id: number; code: string; name: string; workshop: string | null; status: string } | null
  order?: {
    id: number
    code: string
    status: string
    customer_id: number
    customer_name: string | null
    customer_code: string | null
  } | null
  sku?: {
    id: number
    code: string
    name: string
    display_label?: string
    sku_display_name?: string
    product_name?: string
  } | null
  product?: { id: number; code: string; name: string; display_name?: string } | null
  work_order: {
    id: number
    order_id: number
    sku_id: number
    qty: number
    sku_display_label?: string | null
  } | null
}

export type UserOut = {
  id: number
  tenant_id: number
  department_id: number | null
  username: string
  full_name: string | null
  is_active: boolean
  is_superuser: boolean
  created_at: string
  roles: { id: number; code: string; name: string }[]
  department: { id: number; code: string; name: string } | null
}

export const workOrderApi = {
  // 工单
  listWorkOrders(params: any) {
    return http.request<ListResp<WorkOrderOut>>({ url: '/admin/production/work-orders', method: 'GET', params })
  },
  getWorkOrder(id: number) {
    return http.request<WorkOrderDetailOut>({ url: `/admin/production/work-orders/${id}`, method: 'GET' })
  },
  printProductLabels(workOrderId: number, params?: { piece_no_from?: number; piece_no_to?: number }) {
    return http.request<{ html: string; count: number; work_order_id: number }>({
      url: `/admin/production/work-orders/${workOrderId}/print-product-labels`,
      method: 'GET',
      params,
    })
  },

  // 任务
  listTasks(params: any) {
    return http.request<ListResp<TaskOut>>({ url: '/admin/production/tasks', method: 'GET', params })
  },
  assignTask(taskId: number, data: { assigned_user_id: number | null; equipment_id?: number | null }) {
    return http.request<TaskOut>({ url: `/admin/production/tasks/${taskId}/assign`, method: 'POST', data })
  },
  getTaskAssignments(taskId: number) {
    return http.request<{
      task_id: number
      planned_qty: number
      assigned_total_qty: number
      items: TaskAssignmentOut[]
    }>({ url: `/admin/production/tasks/${taskId}/assignments`, method: 'GET' })
  },
  setTaskAssignments(
    taskId: number,
    data: { items: { user_id: number; assigned_qty: number }[]; equipment_id?: number | null },
  ) {
    return http.request<TaskOut>({ url: `/admin/production/tasks/${taskId}/assignments`, method: 'PUT', data })
  },
  listDispatchAssignments(params?: {
    keyword?: string
    order_id?: number
    user_id?: number
    process_id?: number
    offset?: number
    limit?: number
  }) {
    return http.request<{ items: DispatchAssignmentOut[]; total: number }>({
      url: '/admin/production/assignments',
      method: 'GET',
      params,
    })
  },
  getDispatchAssignmentQr(assignmentId: number) {
    return http.request<{ task_code: string; text: string; report_url: string; svg: string }>({
      url: `/admin/production/assignments/${assignmentId}/qr`,
      method: 'GET',
    })
  },
  deleteDispatchAssignment(assignmentId: number) {
    return http.request<void>({ url: `/admin/production/assignments/${assignmentId}`, method: 'DELETE' })
  },
  renderTaskLabel(taskId: number, params?: { template_id?: number; template_code?: string }) {
    return http.request<{ html: string; task_id: number; task_code: string; template_id: number }>({
      url: `/admin/production/tasks/${taskId}/print-label`,
      method: 'GET',
      params,
    })
  },
  exportTaskLabelPdf(taskId: number, params?: { template_id?: number; template_code?: string }) {
    return http.request<{ attachment_id: number; filename: string; url: string }>({
      url: `/admin/production/tasks/${taskId}/print-label-pdf`,
      method: 'GET',
      params,
    })
  },
  renderTaskLabelBatch(data: { task_ids: number[]; template_id?: number; template_code?: string }) {
    return http.request<{ html: string; count: number; template_id: number }>({
      url: '/admin/production/tasks/print-label-batch',
      method: 'POST',
      data,
    })
  },
  listDispatchSkills(params?: any) {
    return http.request<ListResp<{ id: number; code: string; name: string }>>({ url: '/admin/production/tasks/dispatch-skills', method: 'GET', params })
  },
  listDispatchUsers(params: any) {
    return http.request<ListResp<UserOut>>({ url: '/admin/production/tasks/dispatch-users', method: 'GET', params })
  },

  // 用户（派工用）
  listUsers(params: any) {
    return http.request<ListResp<UserOut>>({ url: '/admin/system/users', method: 'GET', params })
  },
}
